backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java

```diff
@@ -1967,97 +1970,144 @@ public void onClick(DialogInterface dialog, int which) {
-							}
-						}
-					}
+	private static void internalAppFolderCleanup(boolean sideUpgrade, File dataDir, String[] filesItenary) {
+		HashSet<String> filesToKeep = new HashSet<>(Arrays.asList(filesItenary));
+
+		File[] extfiles = dataDir.listFiles();
+		if (extfiles == null || extfiles.length == 0) {
+			return;
+		}
+		Log.d(ScummVM.LOG_TAG, "Cleaning up files in " + dataDir.getPath());
+		for (File extfile : extfiles) {
+			String name = extfile.getName();
+
+			if (filesToKeep.contains(name) && sideUpgrade) {
+				continue;
+			}
+			if (extfile.isDirectory()) {
+				// Skip icons, saves
+				if ((name.compareToIgnoreCase("icons") == 0) ||
+					(name.compareToIgnoreCase("saves") == 0)) {
+					continue;
+				}
+				Log.d(ScummVM.LOG_TAG, "Deleting folder:" + extfile.getName());
+				deleteDir(extfile);
+			} else {
+				// Skip scummvm.ini, scummvm.log
+				if ((name.compareToIgnoreCase("scummvm.ini") == 0) ||
+					(name.compareToIgnoreCase("scummvm.log") == 0)) {
+					continue;
+				}
+				Log.d(ScummVM.LOG_TAG, "Deleting file:" + extfile.getName());
+				if (!extfile.delete()) {
+					Log.e(ScummVM.LOG_TAG, "Failed to delete file:" + extfile.getName());
 				}
 			}
 		}
 	}
 
 	// code based on https://stackoverflow.com/a/4530294
 	// Note, the following assumptions are made (since they are true as of yet)
-	// - We don't need to copy (sub)folders
 	// - We copy all the files from our assets (not a subset of them)
 	// Otherwise we would probably need to create a specifically named zip file with the selection of files we'd need to extract to the internal memory
-	private void copyAssetsToInternalMemory(boolean sideUpgrade) {
-		// sideUpgrade is set to true, if we upgrade to the same version -- just check for the files existence before copying
-		if (_actualScummVMDataDir != null) {
-			AssetManager assetManager = getAssets();
-			String[] files = null;
-			try {
-				files = assetManager.list("");
-			} catch (IOException e) {
-				Log.e(ScummVM.LOG_TAG, "Failed to get asset file list.", e);
+	private static boolean extractAssets(boolean sideUpgrade, AssetManager assetManager, String assetDir, File dataDir) {
+		String[] files = null;
+		try {
+			files = assetManager.list(assetDir);
+		} catch (IOException e) {
+			Log.e(ScummVM.LOG_TAG, "Failed to get asset file list.", e);
+		}
+
+		if (files == null || files.length == 0) {
+			// The asset is a file: remove any directory with the same name
+			if (dataDir.isDirectory()) {
+				deleteDir(dataDir);
 			}
+			return false;
+		}
 
-			internalAppFolderCleanup(files, sideUpgrade);
+		// Cleanup old files
+		internalAppFolderCleanup(sideUpgrade, dataDir, files);
 
-			if (files != null) {
-				for (String filename : files) {
-					InputStream in = null;
-					OutputStream out = null;
+		if (!dataDir.exists()) {
+			if (!dataDir.mkdir()) {
+				Log.e(ScummVM.LOG_TAG, "Failed to create directory: " + dataDir.getPath());
+				// There is no point on continuing this
+				return true;
```

Comment:
Shouldn't we return false here, as at this point we weren't able to create the needed directory?

Reply:
We have to return true because true/false doesn't mean success/failure but if the asset is a file (false) or a directory (true)